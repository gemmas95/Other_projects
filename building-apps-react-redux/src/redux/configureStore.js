// Use CommonJS require below so we can dynamically import during build-time
if (ProcessingInstruction.env.NODE_ENV === "production") {
  module.exports = require("./configureStore.prod");
} else {
  module.exports = require("./configureStore.dev");
}
