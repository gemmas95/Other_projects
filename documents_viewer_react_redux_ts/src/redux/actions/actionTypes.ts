import IDocument from "../../models/document.interface";

// Documents actions
export const LOAD_DOCUMENTS: string = "LOAD_DOCUMENTS";
export const CREATE_DOCUMENT_SUCCESS: string = "CREATE_DOCUMENT_SUCCESS";
export const DELETE_DOCUMENT_SUCCESS: string = "DELETE_DOCUMENT_SUCCESS";
export const FILTER_DOCUMENTS_BY_TYPE: string = "FILTER_DOCUMENTS_BY_TYPE";

// Api calls actions
export const LOADING: string = "LOADING";
export const NO_LOADING: string = "NO_LOADING";

export const API_CALL_ERROR: string = "API_CALL_ERROR";

// Interfaces actions
export interface LoadDocumentsAction {
  type: typeof LOAD_DOCUMENTS;
  documents: IDocument[];
}

export interface CreateDocumentSuccessAction {
  type: typeof CREATE_DOCUMENT_SUCCESS;
  document: IDocument;
}

export interface DeleteDocumentSuccessAction {
  type: typeof DELETE_DOCUMENT_SUCCESS;
  documentId: string;
}

export interface FilterDocumentsByType {
  type: typeof FILTER_DOCUMENTS_BY_TYPE;
  typeFilter: string;
}

export interface LoadingAction {
  type: typeof LOADING;
}

export interface NoLoadingAction {
  type: typeof NO_LOADING;
}

export interface ApiCallErrorAction {
  type: typeof API_CALL_ERROR;
}

export type DocumentsActionTypes =
  | LoadDocumentsAction
  | CreateDocumentSuccessAction
  | DeleteDocumentSuccessAction
  | FilterDocumentsByType;

export type ApiCallActionTypes =
  | LoadingAction
  | NoLoadingAction
  | ApiCallErrorAction;

export type AppActions = DocumentsActionTypes | ApiCallActionTypes;
