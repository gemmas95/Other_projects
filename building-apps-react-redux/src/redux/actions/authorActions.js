import * as types from "./actionTypes";
import * as authorApi from "../../api/authorApi";

export function loadAuthorsSuccess(authors) {
  return { type: types.LOAD_AUTHORS_SUCCESS, authors };
}

// THUNK MIDDLEWARE, we don't need to pass dispatch as an argument, redux thunk injects dispacth
export function loadAuthors() {
  return function (dispatch) {
    return authorApi
      .getAuthors()
      .then((authors) => {
        dispatch(loadAuthorsSuccess(authors));
      })
      .catch((error) => {
        throw error;
      });
  };
}
