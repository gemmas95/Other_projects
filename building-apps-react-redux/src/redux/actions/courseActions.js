import * as types from "./actionTypes";
import * as courseApi from "../../api/courseApi";

// Action creator, the only must is to have a type
export function createCourse(course) {
  return { type: types.CREATE_COURSE, course };
}

export function loadCourseSuccess(courses) {
  return { type: types.LOAD_COURSES_SUCCESS, courses };
}

// THUNK MIDDLEWARE, we don't need to pass dispatch as an argument, redux thunk injects dispacth
export function loadCourses() {
  return function (dispatch) {
    return courseApi
      .getCourses()
      .then((courses) => {
        dispatch(loadCourseSuccess(courses));
      })
      .catch((error) => {
        throw error;
      });
  };
}
