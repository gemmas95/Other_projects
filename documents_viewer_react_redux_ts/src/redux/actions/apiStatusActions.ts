import * as types from "./actionTypes";
import { AppActions } from "./actionTypes";

export function beginApiCall(): AppActions {
  return { type: types.LOADING };
}

export function endApiCall(): AppActions {
  return { type: types.NO_LOADING };
}

export function apiCallError(): AppActions {
  return { type: types.API_CALL_ERROR };
}
