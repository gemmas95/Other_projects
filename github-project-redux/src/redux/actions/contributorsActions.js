import * as types from "./actionTypes";
import { getContributors } from "../../api/contributorsApi";
import { apiCallError } from "../actions/apiStatusAction";

export function loadContributorsSuccess(contributors) {
  return { type: types.LOAD_CONTRIBUTORS_SUCCESS, contributors };
}
function loadingCall(commons) {
  return { type: types.LOADING, commons };
}
function loadingSuccess(commons) {
  return { type: types.NO_LOADING, commons };
}

export function loadContributors(data) {
  // console.log("this is DISPACTH CONTRIBUTORSACTIONS", dispatch);
  return function (dispatch) {
    dispatch(loadingCall());
    // dispatch(beginApiCall());

    return getContributors(data)
      .then((response) => response.json())

      .then((contributors) => {
        dispatch(loadContributorsSuccess(contributors));
        dispatch(loadingSuccess());
      })
      .catch((error) => {
        dispatch(apiCallError(error));
        console.error(error);
        throw error;
      });
  };
}

export function sortByLessContributions(contributors) {
  console.log("ai que LLEGAAAA ....", contributors);

  return function (dispatch) {
    dispatch({
      type: types.FILTER_CONTRIBUTORS_BY_LESS_CONTRIBUTIONS,
      contributors,
      /*       contributors: [
        ...contributors.sort((a, b) => {
          console.log("this is a", a, "and this is b", b);
          return a.contributions - b.contributions;
        }),
      ], */
    });
    console.log("ai pollito....", contributors);
  };

  /*  return {
    ...state,
    contributors:
      state.contributors.length > 0
        ? [
            ...state.contributors.sort(
              (a, b) => a.contributions - b.contributions
            ),
          ]
        : state.contributors,
  }; */
}

export function sortByMoreContributions(contributors) {
  return function (dispatch) {
    dispatch({
      type: types.FILTER_CONTRIBUTORS_BY_MORE_CONTRIBUTIONS,
      contributors,
    });
  };
}

export function sortByNameAscending(contributors) {
  return function (dispatch) {
    dispatch({
      type: types.FILTER_CONTRIBUTORS_BY_ASCENDING_NAME,
      contributors,
    });
  };
}

export function sortByNameDescending(contributors) {
  return function (dispatch) {
    dispatch({
      type: types.FILTER_CONTRIBUTORS_BY_DESCENDING_NAME,
      contributors,
    });
  };
}
