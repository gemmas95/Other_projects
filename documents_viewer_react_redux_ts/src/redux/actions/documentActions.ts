import * as types from "./actionTypes";
import { AppActions } from "./actionTypes";
import IDocument from "../../models/document.interface";
import { Dispatch } from "react";

// Imports of fake api calls
import {
  getDocuments,
  saveDocument,
  deletingDocument,
} from "../../api/documentAPI";

import { beginApiCall, apiCallError, endApiCall } from "./apiStatusActions";

// Action types from action creators

export function loadDocumentSuccess(documents: IDocument[]): AppActions {
  return { type: types.LOAD_DOCUMENTS, documents };
}

export function createDocumentSuccess(document: IDocument): AppActions {
  // New documents it's saved in action property document
  return { type: types.CREATE_DOCUMENT_SUCCESS, document };
}

export function removeDocument(documentId: string): AppActions {
  return { type: types.DELETE_DOCUMENT_SUCCESS, documentId };
}

export function filterDocumentsByType(typeFilter: string): AppActions {
  return { type: types.FILTER_DOCUMENTS_BY_TYPE, typeFilter };
}

// Action creators

export function loadDocuments() {
  return async function (dispatch: Dispatch<AppActions>) {
    dispatch(beginApiCall());
    return await getDocuments()
      .then((documents) => {
        dispatch(loadDocumentSuccess(documents));
        dispatch(endApiCall());
      })
      .catch((error) => {
        dispatch(apiCallError());
        console.error(error);
        throw error;
      });
  };
}

// Notifications with toast to improve user experience could be handled here or in the component. I don't want
// to do action creators and the functions that handle them  too verbose so notifications are handled in components
export function createDocument(document: IDocument) {
  return async function (dispatch: Dispatch<AppActions>) {
    dispatch(beginApiCall());
    return await saveDocument(document)
      .then((savedDocument) => {
        dispatch(createDocumentSuccess(savedDocument));
        dispatch(endApiCall());
      })
      .catch((error) => {
        dispatch(apiCallError());
        console.error(error);
        throw error;
      });
  };
}

// Here I delete document optimistically for a better user experience
export function deleteDocument(id: string) {
  return async function (dispatch: Dispatch<AppActions>) {
    dispatch(removeDocument(id));
    return await deletingDocument(id).catch((error) => {
      console.error(error);
      throw error;
    });
  };
}

export function filterDocuments(type: string) {
  return function (dispatch: Dispatch<AppActions>) {
    dispatch(filterDocumentsByType(type));
  };
}
