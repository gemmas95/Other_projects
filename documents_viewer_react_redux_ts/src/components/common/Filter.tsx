import React, { ChangeEvent } from "react";
import { connect } from "react-redux";
import { ThunkDispatch } from "redux-thunk";
import { AppActions } from "../../redux/actions/actionTypes";
import { filterDocuments } from "../../redux/actions/documentActions";
import styled from "styled-components";

// Interfaces
import IDocument from "../../models/document.interface";
import { AppState } from "../../redux/reducers";

const FilterRow = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-around;

  p {
    font-weight: bolder;
  }
`;

interface Props {
  filterDocuments: Function;
  documents: IDocument[];
}

const Filter = (props: Props): JSX.Element => {
  function handleChange(event: ChangeEvent<HTMLSelectElement>) {
    event.preventDefault();
    props.filterDocuments(event.target.value);
  }

  return (
    <FilterRow>
      <p>{props.documents.length} documents found</p>
      <label>
        Filter by
        <select
          data-testid="select"
          onChange={(event) => {
            handleChange(event);
          }}
        >
          <option value="">All</option>
          <option value="simple">Simple</option>
          <option value="custom">Custom</option>
          <option value="advanced">Advanced</option>
        </select>
      </label>
    </FilterRow>
  );
};

function mapStateToProps(state: AppState) {
  return {
    documents: state.documents.visibleDocs,
  };
}

function mapDispatchToProps(dispatch: ThunkDispatch<any, any, AppActions>) {
  return {
    filterDocuments: (type: string) => dispatch(filterDocuments(type)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Filter);
