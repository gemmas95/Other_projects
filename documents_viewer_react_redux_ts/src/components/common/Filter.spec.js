import renderer from "react-test-renderer";
import React from "react";
import Filter from "./Filter";
import { BrowserRouter } from "react-router-dom";

import { Provider } from "react-redux";
import configureStore from "redux-mock-store";

import { act } from "react-dom/test-utils";
import { render } from "@testing-library/react";

import { fireEvent } from "@testing-library/react";

// Action
import { filterDocuments } from "../../redux/actions/documentActions";

const mockStore = configureStore([]);

describe("Filter", () => {
  let container;
  let store;

  let documents;

  beforeEach(() => {
    // To have acces to select element
    container = document.createElement("div");
    document.body.appendChild(container);

    // To wrap component with a Provider and store
    store = mockStore({
      myState: "sample text",
      documents: {
        visibleDocs: [{}],
      },
    });

    store.dispatch = jest.fn();
  });
  afterEach(() => {
    document.body.removeChild(container);
    container = null;
  });
  it("should match snapshot", () => {
    documents = [{}];

    const FilterTree = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <Filter filterDocuments={filterDocuments} documents={documents} />
        </BrowserRouter>{" "}
      </Provider>,
      container
    );
    expect(FilterTree.toJSON()).toMatchSnapshot();
  });

  it("should dispacth filterDocuments", () => {
    act(() => {
      render(
        <Provider store={store}>
          <BrowserRouter>
            <Filter filterDocuments={filterDocuments} documents={documents} />
          </BrowserRouter>
        </Provider>,
        container
      );
    });

    // getByTestId not working so I use querySelector instead
    fireEvent.change(document.querySelector('[data-testid="select"]'), {
      target: { value: "advanced" },
    });

    documents = [{}];

    expect(store.dispatch).toHaveBeenCalledTimes(1);
  });
});
