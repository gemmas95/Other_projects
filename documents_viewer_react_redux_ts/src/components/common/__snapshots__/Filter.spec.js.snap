// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter should match snapshot 1`] = `
Array [
  <div
    className="sc-bdfBwQ cquHKG"
  >
    <p>
      1
       documents found
    </p>
    <label>
      Filter by
      <select
        data-testid="select"
        onChange={[Function]}
      >
        <option
          value=""
        >
          All
        </option>
        <option
          value="simple"
        >
          Simple
        </option>
        <option
          value="custom"
        >
          Custom
        </option>
        <option
          value="advanced"
        >
          Advanced
        </option>
      </select>
    </label>
  </div>,
  " ",
]
`;
